File: results/1.0/ELF/elf-gpt-4o.c
#include <stdio.h>
#include <stdlib.h>
#include <hammer/hammer.h>

// Define parsers for ELF Header fields
HParser *create_elf_header_parser() {
    return h_sequence(
        h_sequence(h_ch('\x7F'), h_ch('E'), h_ch('L'), h_ch('F'), NULL),
        h_uint8(),   // EI_CLASS
        h_uint8(),   // EI_DATA
        h_uint8(),   // EI_VERSION
        h_uint8(),   // EI_OSABI
        h_uint8(),   // EI_ABIVERSION
        h_repeat_n(h_uint8(), 7), // Padding bytes
        h_uint16(),   // e_type
        h_uint16(),   // e_machine
        h_uint32(),   // e_version
        h_uint32(),   // e_entry
        h_uint32(),   // e_phoff
        h_uint32(),   // e_shoff
        h_uint32(),   // e_flags
        h_uint16(),   // e_ehsize
        h_uint16(),   // e_phentsize
        h_uint16(),   // e_phnum
        h_uint16(),   // e_shentsize
        h_uint16(),   // e_shnum
        h_uint16(),   // e_shstrndx
        NULL
    );
}

// Define parsers for Program Header Table
HParser *create_program_header_parser() {
    return h_sequence(
        h_uint32(),   // p_type
        h_uint32(),   // p_offset
        h_uint32(),   // p_vaddr
        h_uint32(),   // p_paddr
        h_uint32(),   // p_filesz
        h_uint32(),   // p_memsz
        h_uint32(),   // p_flags
        h_uint32(),   // p_align
        NULL
    );
}

// Define parsers for Section Header Table
HParser *create_section_header_parser() {
    return h_sequence(
        h_uint32(),   // sh_name
        h_uint32(),   // sh_type
        h_uint32(),   // sh_flags
        h_uint32(),   // sh_addr
        h_uint32(),   // sh_offset
        h_uint32(),   // sh_size
        h_uint32(),   // sh_link
        h_uint32(),   // sh_info
        h_uint32(),   // sh_addralign
        h_uint32(),   // sh_entsize
        NULL
    );
}

// Function to parse ELF file
void parse_elf(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Failed to open file");
        exit(EXIT_FAILURE);
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    unsigned char *buffer = (unsigned char *)malloc(file_size);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(file);
        exit(EXIT_FAILURE);
    }

    fread(buffer, 1, file_size, file);
    fclose(file);

    HParser *elf_header_parser = create_elf_header_parser();
    HParseResult *result = h_parse(elf_header_parser, buffer, file_size);
    if (result) {
        printf("ELF Header parsed successfully.\n");
        h_parse_result_free(result);
    } else {
        printf("Failed to parse ELF Header.\n");
    }

    free(buffer);
    h_parser_free(elf_header_parser);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <elf-file>\n", argv[0]);
        return EXIT_FAILURE;
    }

    parse_elf(argv[1]);

    return EXIT_SUCCESS;
}

File: results/1.0/ELF/elf-gpt-4o.rs
use std::fs::File;
use std::io::{self, Read};
use std::path::Path;
use std::env;
use nom::{
    IResult,
    bytes::complete::{tag, take},
    number::complete::{le_u16, le_u32, le_u64, u8},
    combinator::{map, map_res},
    sequence::{tuple},
    multi::count,
};

#[derive(Debug)]
struct ElfHeader {
    e_ident: [u8; 16],
    e_type: u16,
    e_machine: u16,
    e_version: u32,
    e_entry: u64,
    e_phoff: u64,
    e_shoff: u64,
    e_flags: u32,
    e_ehsize: u16,
    e_phentsize: u16,
    e_phnum: u16,
    e_shentsize: u16,
    e_shnum: u16,
    e_shstrndx: u16,
}

fn parse_ident(input: &[u8]) -> IResult<&[u8], [u8; 16]> {
    map(take(16usize), |bytes: &[u8]| {
        let mut ident = [0; 16];
        ident.copy_from_slice(bytes);
        ident
    })(input)
}

fn parse_elf_header(input: &[u8]) -> IResult<&[u8], ElfHeader> {
    map(tuple((
        parse_ident,
        le_u16, le_u16, le_u32,
        map(le_u64, |v| if v == 0 { 0 } else { v }),
        le_u64, le_u64,
        le_u32, le_u16,
        le_u16, le_u16,
        le_u16, le_u16,
        le_u16
    )), |(
        e_ident, e_type, e_machine, e_version, e_entry,
        e_phoff, e_shoff, e_flags, e_ehsize,
        e_phentsize, e_phnum, e_shentsize, e_shnum, e_shstrndx
    )| ElfHeader {
        e_ident,
        e_type,
        e_machine,
        e_version,
        e_entry,
        e_phoff,
        e_shoff,
        e_flags,
        e_ehsize,
        e_phentsize,
        e_phnum,
        e_shentsize,
        e_shnum,
        e_shstrndx,
    })(input)
}

fn read_file<P: AsRef<Path>>(path: P) -> io::Result<Vec<u8>> {
    let mut file = File::open(path)?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;
    Ok(buffer)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <elf-file>", args[0]);
        return;
    }

    let filename = &args[1];
    let data = read_file(filename).expect("Failed to read file");

    match parse_elf_header(&data) {
        Ok((_, elf_header)) => {
            println!("{:?}", elf_header);
        },
        Err(e) => {
            eprintln!("Failed to parse ELF header: {:?}", e);
        }
    }
}